Add Bookmarks API to the browser namespace

akutira-umich opened the pull request:
#### bbf50b748dba84f51f7002498e72f8db18bc1f6c
<pre>
Add Bookmarks API to the browser namespace
<a href="https://rdar.apple.com/152868076">rdar://152868076</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294224">https://bugs.webkit.org/show_bug.cgi?id=294224</a>

Reviewed by Brian Weinstein and Timothy Hatcher.

Add Bookmarks API to the browser namespace. Files included to check for bookmark
permissions in the manifests and have test files.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Extensions/WebExtensionPermission.cpp:
(WebKit::WebExtensionPermission::bookmarks):
* Source/WebKit/Shared/Extensions/WebExtensionPermission.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPermission.mm: Gives access to
Bookmark API if bookmarks are enabled in the mm file
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPermissionPrivate.h:
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::supportedPermissions):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed): Checks for bookmarks in the manifest
(WebKit::WebExtensionAPINamespace::bookmarks):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h: Copied from Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPermissionPrivate.h.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIBookmarks.idl: Copied from Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPermissionPrivate.h.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm: Added.
(TestWebKitAPI::WKWebExtensionAPIBookmarks::WKWebExtensionAPIBookmarks):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::getManagerFor):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, APISUnavailableWhenManifestDoesNotRequest)):
Tests when the manifest does not includes the bookmarks permissions
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, APIAvailableWhenManifestRequests)): Tests when the manifest includes the bookmarks permissions
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
(TestWebKitAPI::WKWebExtensionAPISidebar::WKWebExtensionAPISidebar):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm: Included import to fix building issues

Canonical link: <a href="https://commits.webkit.org/296338@main">https://commits.webkit.org/296338@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42492c91fb1cf56886cd322e9076da9d31d4dea7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108208 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27870 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18291 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113418 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110171 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28569 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36421 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82160 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111156 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22637 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97469 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62591 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22053 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15605 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58153 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92003 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15661 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116544 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35272 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25973 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91187 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35646 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93745 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90982 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/23183 "The change is no longer eligible for processing.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35873 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13631 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31054 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17480 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35171 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40726 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34907 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38261 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36572 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/763710d97fa09e5bea9fac09fa57ffdb7bcf61bf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46737

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
```diff
@@ -88,6 +90,11 @@
+#if ENABLE(WK_WEB_EXTENSIONS_BOOKMARKS)
+    if (name == "bookmarks"_s)
+        return page->corePage()->settings().webExtensionBookmarksEnabled() && extensionContext().hasPermission("bookmarks"_s);
+
```
nit: can drop this newline

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
```diff
@@ -266,6 +273,18 @@
+#if ENABLE(WK_WEB_EXTENSIONS_BOOKMARKS)
+WebExtensionAPIBookmarks& WebExtensionAPINamespace::bookmarks()
+{
+    // Documentation: https://developer.chrome.com/docs/extensions/reference/api/bookmarks
```
we use mdn for documentation links
https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/bookmarks

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -0,0 +1,47 @@
+
+
+public:
+#if PLATFORM(COCOA)
```
i think you can drop this #if #endif and add it when you add the method definitions in the follow up PR

AriYoung00 replied:
Ditto for `public:`

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -0,0 +1,154 @@
+    WKWebExtensionAPIBookmarks()
+    {
+        NSLog(@"[DEBUG] WKWebExtensionAPIBookmarks constructor: START");
+        bookmarkConfig = WKWebExtensionControllerConfiguration.defaultConfiguration;
```
i see that Sidebar also used the defaultConfiguration, but i _think_ you'd want a temporary configuration here? cc @xeenon

kiaraarose replied:
i see the sidebar test also using the defaultConfiguration. should these use the temporary or non persistent one? more of a question for @xeenon

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -0,0 +1,154 @@
+protected:
+    WKWebExtensionAPIBookmarks()
+    {
+        NSLog(@"[DEBUG] WKWebExtensionAPIBookmarks constructor: START");
```
is this log meant to be here?

kiaraarose replied:
i see another test using `DEBUG_LOG`. i think that would be better to use

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -0,0 +1,154 @@
+        for (_WKFeature *feature in [WKPreferences _features]) {
+            if ([feature.key isEqualToString:@"WebExtensionBookmarksEnabled"]) {
+                targetFeature = feature;
+                // Attempt to enable
```
nit: i don't think this comment is needed

AriYoung00 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPermission.h:
```diff
@@ -90,6 +90,10 @@ WK_EXTERN WKWebExtensionPermission const WKWebExtensionPermissionUnlimitedStorag
 
+/*! @abstract The `bookmarks` permission requests access to the `browser.bookmarks` APIs. */
+WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA))
+WK_EXTERN WKWebExtensionPermission const WKWebExtensionPermissionBookmarks NS_SWIFT_NONISOLATED;
```
I believe this should go in `WKWebExtensionPermissionPrivate.h` since I don't think we want it to be public API yet.

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -0,0 +1,154 @@
+
+TEST_F(WKWebExtensionAPIBookmarks, APISUnavailableWhenManifestDoesNotRequest)
+{
+    NSLog(@"[DEBUG TEST] APISUnavailableWhenManifestDoesNotRequest: START");
```
i don't think this log adds much

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -0,0 +1,154 @@
+
+TEST_F(WKWebExtensionAPIBookmarks, APIAvailableWhenManifestRequests)
+{
+    NSLog(@"[DEBUG TEST] APIAvailableWhenManifestRequests: START");
```
ditto

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -0,0 +1,154 @@
+
+#pragma mark - Common Tests
+
+TEST_F(WKWebExtensionAPIBookmarks, APISUnavailableWhenManifestDoesNotRequest)
```
i think this test name is a bit misleading. the API shouldn't be available just because the manifest requests it. granting the permission on line 146 is what makes it available

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -0,0 +1,154 @@
+
+#pragma mark - Constants
+
+static constexpr auto *bookmarkOffManifest = @{
```
i think you can remove this manifest and just have the one below

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -0,0 +1,154 @@
+    },
+};
+
+static auto *bookmarkOnManifest = @{
```
same comment here on the name being misleading. i think this could just be `bookmarksManifest`

kiaraarose commented:
there are two commits in this PR. WebKit guidance is to use one per PR, so you should quash them

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -0,0 +1,155 @@
+    },
+};
+
+
```
nit: can drop extra newline

kiaraarose reviewed (changes requested):
Left some comments. Major requests would be to squash the commits and move the bookmarks permission into the Private header

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
```diff
@@ -88,6 +90,11 @@
 #endif // ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
 
+#if ENABLE(WK_WEB_EXTENSIONS_BOOKMARKS)
+    if (name == "bookmarks"_s)
```
should move this up after "action" to preserve the order

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
```diff
@@ -30,6 +30,7 @@
 #include "JSWebExtensionAPINamespace.h"
 #include "WebExtensionAPIAction.h"
 #include "WebExtensionAPIAlarms.h"
+#include "WebExtensionAPIBookmarks.h"
```
can this import be guarded?

akutira-umich replied:
Yes I believe I can guard this with #if ENABLE(WK_WEB_EXTENSIONS_BOOKMARKS)

kiaraarose commented:
I think a better title for this PR would be "Add Bookmarks API to the browser namespace"

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a4341a6685276010c13a961e351353ac378fbcab. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46737

AriYoung00 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -0,0 +1,145 @@
+        }
+
+        if (targetFeature) {
+            NSLog(@"[DEBUG_LOG] WebExtensionBookmarksEnabled feature flag was targeted. Attempted to enable. Current state: %s",
```
Is this logging (and the associated code above) necessary?

akutira-umich replied:
So the point of this was to make sure the feature flag was enabling properly. It was helpful in debugging but now if it doesn't need to be there I can remove it

kiaraarose commented on Tools/TestWebKitAPI/SourcesCocoa.txt:
```diff
@@ -343,6 +343,7 @@ Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm
 Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm
 Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm
 Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm
+Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm
```
ditto

AriYoung00 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    WKWebExtensionControllerConfiguration *bookmarkConfig;
+    RetainPtr<WKWebExtension> extension;
```
Looks like this is unused right? If so, it should probably be removed

akutira-umich committed:
Add Bookmarks API to the browser namespace
rdar://152868076
https://bugs.webkit.org/show_bug.cgi?id=294224

Reviewed by Brian Weinstein and Timothy Hatcher.

Add Bookmarks API to the browser namespace. Files included to check for bookmark
permissions in the manifests and have test files.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Extensions/WebExtensionPermission.cpp:
(WebKit::WebExtensionPermission::bookmarks):
* Source/WebKit/Shared/Extensions/WebExtensionPermission.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPermission.mm: Gives access to
Bookmark API if bookmarks are enabled in the mm file
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPermissionPrivate.h:
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::supportedPermissions):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed): Checks for bookmarks in the manifest
(WebKit::WebExtensionAPINamespace::bookmarks):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h: Copied from Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPermissionPrivate.h.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIBookmarks.idl: Copied from Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPermissionPrivate.h.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm: Added.
(TestWebKitAPI::WKWebExtensionAPIBookmarks::WKWebExtensionAPIBookmarks):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::getManagerFor):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, APISUnavailableWhenManifestDoesNotRequest)):
Tests when the manifest does not includes the bookmarks permissions
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, APIAvailableWhenManifestRequests)): Tests when the manifest includes the bookmarks permissions
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
(TestWebKitAPI::WKWebExtensionAPISidebar::WKWebExtensionAPISidebar):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm: Included import to fix building issues

Canonical link: https://commits.webkit.org/296338@main
